### DISCUSSION TITLE
ConjugateHeatTransfer with different element types

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#23846] [Q&A General] [answered]
Hello:

In the attached, there are 2 blocks with an internal sideset between them.  In block 1, the variable T1 is a "constant monomial".  In block 2, T2 is a "first Lagrange".  I am trying to use the ConjugateHeatTransfer interface kernel to pass heat from one to the other.

I get nonconvergence  of the linear solve:

Time Step 1, time = 0.05, dt = 0.05
 0 Nonlinear |R| = 3.162278e+02
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT 
Nonlinear solve did not converge due to DIVERGED_LOCAL_MIN iterations 0

Any ideas on how to couple these 2 types of FEM variables?

Thanks,
  Joe Kelly

[InterfaceTester_4.i.zip](https://github.com/idaholab/moose/files/11072033/InterfaceTester_4.i.zip)

**Comment by GiudGiud** [comment]
Hi Joe

I ll check when I get back but I don’t think we can use regular heat conduction kernels for constant monomials. I think it should be either DG or finite volume kernels to make it work 

Guillaume

**Reply by GiudGiud** [reply]
Yeah I've been trying a few things and the formulation in the Diffusion kernel isnt meant for variables that have a flat shape function, the contribution to the residual is just 0.

What numerical scheme do you want to be using ultimately?

**Reply by joe61vette** [reply]
Thanks for looking at this Guillaume.  I should have figured out that I would need a DGDiffusion kernel.  I am trying out a hybrid rDG/CG scheme. Some simple problems seem to work very well. I would like to have a fluid temperature that is DG but couples to a solid temperature that is CFEM (first Lagrange).  I had a problem with that but will try again later this week.  If I can get that to work, then I will still need a ConjugateHeatTransfer interface kernel that works for one variable being DG and the other CFEM.

**Reply by GiudGiud** [reply]
This seems to do what you want?

@lindsayad for double check please. The interface works, but the solution isnt matching the expected 1D profile (too high on the left). Do you have guidance on setting the DG parameters for coarse mesh diffusion?

```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    nx = 10
    ny = 10
    dim = 2
  []
  [block1]
    type = SubdomainBoundingBoxGenerator
    block_id = 1
    bottom_left = '0 0 0'
    top_right = '0.5 1 0'
    input = gen
  []
  [block2]
    type = SubdomainBoundingBoxGenerator
    block_id = 2
    bottom_left = '0.5 0 0'
    top_right = '1 1 0'
    input = block1
  []
  [sideset]
    input = block2
    type = SideSetsBetweenSubdomainsGenerator
    new_boundary = 'int_ss'
    primary_block = '2'
    paired_block = '1'
  []
[]

[Problem]
  kernel_coverage_check = false
[]

[Variables]
  [T1]
    order = CONSTANT
    family = MONOMIAL
    block = '1'
  []
  [T2]
    order = FIRST
    family = LAGRANGE
    block = '2'
  []
[]

[ICs]
  [t1]
    type = FunctionIC
    variable = T1
    function = '2 + x +y'
  []
[]

[Kernels]
  [cond1]
    type = HeatConduction
    variable = T1
    block = '1'
  []
  [cond2]
    type = HeatConduction
    variable = T2
    block = '2'
  []
[]

[DGKernels]
  active = 'dg_diff'

  [./dg_diff]
    type = DGDiffusion
    variable = T1
    epsilon = -1
    sigma = 6
  [../]
[]



[InterfaceKernels]
  [intf1]
    type = ConjugateHeatTransfer
    variable = T2
    T_fluid = T2
    neighbor_var = T1
    boundary = int_ss
    htc = 'alpha_wall'
  []
[]

[BCs]
  [left_flux]
    type = NeumannBC
    value = 1000
    variable = T1
    boundary = left
  []
  [right_temp]
    type = DirichletBC
    value = 0
    variable = T2
    boundary = right
  []
[]

[Materials]
  [thermal_cond]
    type = GenericConstantMaterial
    prop_names = 'thermal_conductivity'
    prop_values = '10'
  []
  [htc]
    type = ADGenericConstantMaterial
    prop_names = 'alpha_wall'
    prop_values = '1000'
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'

  nl_rel_tol = 1e-10
  nl_abs_tol = 1e-10

  # petsc_options = '-pc_svd_monitor'
  # petsc_options_iname = '-pc_type'
  # petsc_options_value = 'svd'

  dt = 0.05
  num_steps = 1
[]

[Outputs]
  print_linear_residuals = false
  exodus = true
[]
```

**Reply by joe61vette** [reply]
Yes!  That is exactly what I want.  I did try this but didn't have the DGDiffusion kernel, so (if I remember right) I got non-convergence due to the ZEROPIVOT. That should nave given me a clue.

**Reply by lindsayad** [reply]
I think that you need at least `order = FIRST` in order to get an accurate representation of the diffusion operator. If you look at `DGDiffusion` you will see that any term involving the diffusion coefficient will disappear if the elemental values of `_grad_u` and `_grad_test` are zero. Essentially all you have when `order = CONSTANT` is a penalty term for jumps in the solution. You can try running our `2d_diffusion_dg_test.i` with `order = CONSTANT` and you can see that the error asymptotically is approaching a constant (the error does not decrease linearly as you refine the mesh)
```
Postprocessor Values:
+----------------+----------------+----------------+----------------+
| time           | dofs           | h              | l2_err         |
+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   4.000000e+00 |   7.071068e-01 |   4.293667e-02 |
|   2.000000e+00 |   1.600000e+01 |   3.535534e-01 |   3.182482e-02 |
|   3.000000e+00 |   6.400000e+01 |   1.767767e-01 |   2.031878e-02 |
|   4.000000e+00 |   2.560000e+02 |   8.838835e-02 |   1.296973e-02 |
|   5.000000e+00 |   1.024000e+03 |   4.419417e-02 |   9.107862e-03 |
|   6.000000e+00 |   4.096000e+03 |   2.209709e-02 |   7.314127e-03 |
|   7.000000e+00 |   1.638400e+04 |   1.104854e-02 |   6.529575e-03 |
+----------------+----------------+----------------+----------------+
```

**Reply by lindsayad** [reply, marked as answer]
@joe61vette are you trying to develop arbitrarily high order rDG? If you just want second order accurate rDG, then that is what MOOSE finite volume is: P0P1 rDG.

**Reply by joe61vette** [reply]
Hi Alex: this is related to ASP.  I use DG for all advected quantities except for the liquid mass equation that is solved for pressure.  That is CG.  Basically the hybrid rDG/CG scheme you suggested in order to bypass the difficulty associated with a two-phase Rhie-Chow interpolation.  Many simple test cases run very well (I am using upwind).  The AD system is truly awesome by the way.  

Approach has difficulties when full 6 or 9 equation model is solved when the liquid phase doesn't exist. I plan (hope) to revisit once a segregated solver is available in MOOSE. That would allow me to solve the phasic momentum equations together first.  Then the mass and energy equations would not see velocity as a non-linear variable but would see it as an AD value with associated derivatives (e.g., pressure).

The above has to do with the boundary between a fluid region and a solid wall.  I would like to take advantage of CFEM for solid heat conduction.

Thanks for all the help,
  Joe
